packetbeat/npcap/npcap.go

```diff
@@ -0,0 +1,284 @@
+//  }
+//
+// The URL field will point to the location of anb Npcap installer.
+package npcap
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"runtime"
+	"strings"
+
+	"github.com/google/gopacket/pcap"
+	"golang.org/x/mod/semver"
+
+	"github.com/elastic/beats/v7/libbeat/logp"
+)
+
+// RegistryNotFound is returned by CurrentVersion if the network request
+// returns a 404 status code.
+var RegistryNotFound = errors.New("npcap: registry not found")
+
+var (
+	// Installer holds the embedded installer when run with x-pack.
+	Installer []byte
+
+	// EmbeddedInstallerVersion holds the version of the embedded installer.
+	EmbeddedInstallerVersion string
+)
+
+// Fetch downloads the Npcap installer, writes the content to the given filepath
+// and returns the sha256 hash of the downloaded object. If the registry is not
+// found RegistryNotFound is returned.
+func CurrentVersion(ctx context.Context, log *logp.Logger, registry string) (version, url, hash string, err error) {
+	if runtime.GOOS != "windows" {
+		return "", "", "", errors.New("npcap: called Fetch on non-Windows platform")
+	}
+	return currentVersion(ctx, log, registry)
+}
+
+func currentVersion(ctx context.Context, log *logp.Logger, registry string) (version, url, hash string, err error) {
+	req, err := http.NewRequestWithContext(ctx, "GET", registry, nil)
+	if err != nil {
+		return "", "", "", err
+	}
+
+	var client http.Client
+	res, err := client.Do(req)
+	if err != nil {
+		return "", "", "", err
+	}
+	defer res.Body.Close()
+
+	b, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		log.Errorf("failed to read the error response body: %v", err)
+	}
+	b = bytes.TrimSpace(b)
+	if len(b) == 0 {
+		// Give a meaningful error message and make json error if we have no body.
+		b = []byte("empty")
+	}
+	if res.StatusCode != http.StatusOK {
+		if res.StatusCode == http.StatusNotFound {
+			return "", "", "", fmt.Errorf("%w: %s", RegistryNotFound, registry)
+		}
+		if registry == "" {
+			registry = `""`
+		}
+		return "", "", "", fmt.Errorf("npcap: failed to fetch version info at %s, status: %d, message: %s", registry, res.StatusCode, b)
+	}
+
+	var info struct {
+		Version string
+		URL     string
+		Hash    string
+	}
+	err = json.Unmarshal(b, &info)
+	if err != nil {
+		return "", "", "", fmt.Errorf("%w: %#q", err, b)
+	}
+
+	return info.Version, info.URL, info.Hash, nil
+}
+
+// Fetch downloads the Npcap installer, writes the content to the given filepath
+// and returns the sha256 hash of the downloaded object.
+func Fetch(ctx context.Context, log *logp.Logger, url, path string) (hash string, err error) {
+	if runtime.GOOS != "windows" {
+		return "", errors.New("npcap: called Fetch on non-Windows platform")
+	}
+	return fetch(ctx, log, url, path)
+}
+
+func fetch(ctx context.Context, log *logp.Logger, url, path string) (hash string, err error) {
```

Comment:
Having a config option that can be remotely controlled and causes Packetbeat to execute a user-specified binary (from file or download) adds a new threat vector.

Elastic Agent allows downloads of the Beats, but they are required to have a GPG signature from Elastic so this gives us some level of protection.

If we want to allow the user to specify an installer binary location then I think we need to have some checks on the binary before executing it. Like perhaps we can validate that the npcap installer is signed by npcap before executing it. Otherwise I would forgo the configurability of the npcap installer location until there's a good defense mechanism.

Reply:
The intention here is that this approach is only used for distribution within an org. I can make that clearer in the config.

Reply:
Backed out.